iothub/device/src/Transport/AmqpIoT/AmqpIoTReceivingLink.cs

```diff
@@ -258,88 +258,104 @@ private void DisposeDelivery(AmqpMessage amqpMessage, bool settled, Accepted acc
                 string correlationId = amqpMessage.Properties?.CorrelationId?.ToString();
+                int status = getStatus(amqpMessage);
 
-                if (!VerifyResponseMessage(amqpMessage))
+                if (status >= 400)
                 {
+                    // Handle failures
                     _onDesiredPropertyReceived.Invoke(null, correlationId, null);
-                }
-
-                Twin twin = null;
-                TwinCollection twinProperties = null;
+                    string error = null;
+                    using (StreamReader reader = new StreamReader(amqpMessage.BodyStream, System.Text.Encoding.UTF8))
+                    {
+                        error = reader.ReadToEnd();
+                    }
 
-                if (correlationId != null)
+                    if (status >= 500 || status == 429 || status == 408)
+                    {
+                        throw new IotHubException(error, true);
+                    }
+                    else
+                    {
+                        throw new IotHubException(error, false);
+                    }
+                }
+                else
                 {
-                    if (amqpMessage.BodyStream != null)
+                    Twin twin = null;
+                    TwinCollection twinProperties = null;
+
+                    if (correlationId == null)
                     {
-                        // This a result of a GET TWIN so return (set) the full twin
+                        // Here we are getting desired property update notifications and want to handle it first
+                        using (StreamReader reader = new StreamReader(amqpMessage.BodyStream, System.Text.Encoding.UTF8))
+                        {
+                            string patch = reader.ReadToEnd();
+                            twinProperties = JsonConvert.DeserializeObject<TwinCollection>(patch);
+                        }
+                    }
+                    else if (correlationId.StartsWith(AmqpTwinMessageType.Get.ToString(), StringComparison.OrdinalIgnoreCase))
+                    {
+                        // This a response of a GET TWIN so return (set) the full twin
                         using (StreamReader reader = new StreamReader(amqpMessage.BodyStream, System.Text.Encoding.UTF8))
                         {
                             string body = reader.ReadToEnd();
                             var properties = JsonConvert.DeserializeObject<TwinProperties>(body);
                             twin = new Twin(properties);
                         }
                     }
-                    else
+                    else if (correlationId.StartsWith(AmqpTwinMessageType.Patch.ToString(), StringComparison.OrdinalIgnoreCase))
+                    {
+                        // This can be used to coorelate success response with Updating reported properties
```

Comment:
so no action here? if so, do we need this else block?

Reply:
No action here yet. If we go with implementing request response pattern for patch this will be where we'll add code and hence kept the comment.

Reply:
this could use a log statement here

Reply:
I don't see a need for log statement here. we are also not logging for other success cases and really to the current implementation of updating reported properties it is not request response style.

Reply:
Then I think it deserves a comment saying that.

Reply:
Are you suggesting to update comment to say "updating reported property is not a request response style implementation "?

Reply:
if updating reported property isn't waiting for the service response, then which call is returning the exception thrown here: https://github.com/Azure/azure-iot-sdk-csharp/blob/8a3cd019acbe2fd4eebc568e76f7e9d01c1fc702/iothub/device/src/Transport/AmqpIoT/AmqpIoTReceivingLink.cs#L287-L294

Reply:
> this could use a log statement here

Or did you mean that it will be implemented in a following PR? If you are implementing in a following PR then yeah, no need to add a log statement for the interim.

Reply:
Exception is thrown for any twin message we get over amqp where a failure code is returned in message Annotation

Reply:
Sure, but as a caller, which operation/ API call would I need to surround with a try-catch to handle this exception?

Reply:
It should be every twin operation be it get/patch/put

Reply:
So `UpdateReportedPropertiesAsync` does wait for the response from the service (containing either a 204 or error code) before completing, right?

Reply:
I mean for every twin response we are handling exception. We can chat if this is confusing offline.

Reply:
update reported properties is not a request response style so no to that. It's confusing I can totally understand.

Howerver I'll be adding a log statement here.

Reply:
if the `UpdateReportedPropertiesAsync` call isn't waiting for the service response, then I still don't understand where the try-catch block would be for catching the above exception. Sure, ping me when you're available, we can chat offline.

Reply:
No it won't. try-catch block would not be catching exception in case of patch. OR in other words not until we change implementation

Reply:
As we discussed, we don't want to throw exceptions that the caller cannot catch and that will cause the application to crash; we should add a check that the response is for a "get" call and then throw the exception, and we can add the same behavior for "patch/put" in the next PR.